### ISSUE TITLE
Support OSX

### ISSUE THREAD

**Issue by Coder206** [jankeromnes/janitor#23]
It would be great if we could run a noVNC of an OSX machine.

**Comment by jankeromnes** [comment]
Thanks for the suggestion! Do you have any specific use cases or killer features in mind for OSX machines on Janitor?

**Comment by Coder206** [comment]
@jankeromnes As a contributor to the Servo project, we often need to ask key people running OSX on their machines to verify things. Having support for OSX machines would eliminate this. 

Also, the WebKit project would benefit from the Janitor project.

To add, staying on the cutting edge and allowing OSX machines running beta versions of OSX would also be beneficial.

**Comment by Coder206** [comment]
@jankeromnes Can we make an OS wishlist like the #24?

**Comment by EtienneWan** [comment]
Apart of running HackOS inside Docker (to fit into Janitor architecture) (is that even possible ? Some answers from StackOverflow are old but say no, it isn't), the only other possibility I see is to rent a macOS VM (mainly to have XCode).
Some services exist but about pricing, it's another story.

@Coder206 So in fact, an OS wishlist would be limited to OSes with Linux kernel, because they have to run on top of Docker.

**Comment by jankeromnes** [comment]
I'm not sure an OS wishlist would be that useful at this point, because there aren't many OSes out there, and supporting them is tricky enough to justify dedicated issues.

Etienne is right about OSX and Windows not quite fitting into Janitor's Docker container architecture, which is the "magic" piece of technology that allows very efficient scaling (project environments can be as large as 30 GB, but with Docker's copy-on-write feature, you can allow 1000 developers to create many containers very cheaply). Supporting 1000 developers with 30 GB VMs each is another story.

While supporting non-Dockerized OSes will be tricky and expensive, I agree that it's valuable. The Janitor can support non-Docker back-ends, like VMs or even dedicated OSX/Windows servers, but we're not yet able to scale these efficiently and we need to figure out the costs. Maybe we won't have to do that work ourselves though, as services like [SauceLabs](https://saucelabs.com/platforms) already know how to run simulators and dedicated servers/devices efficiently (SauceLabs is already integrated with Cloud9, and they're free for Open Source. They focus mostly on cross-browser web testing, but I believe some companies can rent their simulators/servers for other purposes).

**Comment by Coder206** [comment]
@jankeromnes @EtienneWan I see your point, I understand!

**Comment by Coder206** [comment]
@jankeromnes @EtienneWan I was looking into this. I am trying to understand the difference between Docker for OSX and why it will not work for Janitor. 

Despite this, I found something you might want to see, it's called boot2docker (http://boot2docker.io/). It's a lightweight VM for combining Windows and OSX with Docker compatibilities that seem to solve the challenges described above.

**Comment by EtienneWan** [comment]
Boot2docker is a Linux VM which contains Linux binaries available for all Linux systems that are installed on top of B2D.
Because macOS, or even BSD systems are based on UNIX kernel, theyccan't run on top of B2D.

**Comment by Coder206** [comment]
@EtienneWan If I understand correctly, it is just a boot2docker is just a Linux distro running on an OSX machine?

**Comment by EtienneWan** [comment]
In order to be more understandable : the server "already in production" for Janitor is running with a Linux system and a container system (Docker). This is the lower level of the architecture and won't change.

With Docker, you can install and run multiple Linux systems in parallels and only them (because Docker is designed this way, to be simple). Instead of using Vagrant (which could have given the possibility to install macOS or Windows in VMs, and to offer similar results), Docker is the only option to have many OS in parrallel without burning huge anmount of HDD.

**Comment by EtienneWan** [comment]
@Coder206 Exactly. It gives the basis of Linux to run some more Linux on top of it, no matter the host is Windows or macOS.

**Comment by Coder206** [comment]
@EtienneWan I understand! :) Thank you so much.

**Comment by notriddle-cerebrum** [comment]
COW disk images wouldn't work? I assume the Genuine Windows checker would go off if you did that...

**Comment by Coder206** [comment]
After reading this article (http://arstechnica.com/apple/2016/06/digging-into-the-dev-documentation-for-apfs-apples-new-file-system/), I am under the impression that Apple's new file system may allow having many portions of the OS running in parallel without additional storage. 

Correct me if I read that wrong.

**Comment by jankeromnes** [comment]
Random note, this might help with hosting: https://xcloud.me/
